### PR TITLE
etcdserver: support MemberAdd for learner

### PR DESCRIPTION
**Pull request by jingyih**
Support MemberAdd for learner on server side.

1) Added IsLearner field to etcdserver internal Member type.
2) Routed learner MemberAdd request from server API to Raft.
3) Apply learner MemberAdd result to server after the request is passed through Raft.

cc @WIZARD-CXY @xiang90 @jpbetz @gyuho

### CODE REVIEW

File: etcdserver/api/membership/cluster_test.go [hunk 1]
```diff
@@ -461,7 +461,7 @@ func TestClusterAddMember(t *testing.T) {
 			Params: []interface{}{
 				path.Join(StoreMembersPrefix, "1", "raftAttributes"),
 				false,
-				`{"peerURLs":null}`,
+				`{"peerURLs":null,"isLearner":false}`,
```
Comment (jpbetz) [inline etcdserver/api/membership/cluster_test.go]: If isLeadner is absent, does it default to false? Test both cases (where "isLearner" is false, and when it's absent)?

File: etcdserver/api/membership/cluster.go [hunk 2]
```diff
@@ -62,7 +62,7 @@ type RaftCluster struct {
 func NewClusterFromURLsMap(lg *zap.Logger, token string, urlsmap types.URLsMap) (*RaftCluster, error) {
 	c := NewCluster(lg, token)
 	for name, urls := range urlsmap {
-		m := NewMember(name, urls, token, nil)
+		m := NewMember(name, urls, token, nil, false) // does not support adding learner via URLsMap
```
Comment (jpbetz) [inline etcdserver/api/membership/cluster.go]: Add godoc to function and document that learner support is not available in the godoc?

File: etcdserver/api/membership/member.go [hunk 3]
```diff
@@ -35,6 +35,8 @@ type RaftAttributes struct {
 	// PeerURLs is the list of peers in the raft cluster.
 	// TODO(philips): ensure these are URLs
 	PeerURLs []string `json:"peerURLs"`
+	// IsLearner indicates if the member is raft learner.
+	IsLearner bool `json:"isLearner"`
```
Comment (jpbetz) [inline etcdserver/api/membership/member.go]: Do we need to allow this to be absent for backward compatibility?

File: etcdserver/api/membership/member.go [hunk 4]
```diff
@@ -51,10 +53,13 @@ type Member struct {
 
 // NewMember creates a Member without an ID and generates one based on the
 // cluster name, peer URLs, and time. This is used for bootstrapping/adding new member.
-func NewMember(name string, peerURLs types.URLs, clusterName string, now *time.Time) *Member {
+func NewMember(name string, peerURLs types.URLs, clusterName string, now *time.Time, isLearner bool) *Member {
```
Comment (jpbetz) [inline etcdserver/api/membership/member.go:56]: Feels like at some point we should replace the positional parameters with a options struct that is passed in so we don't have to go change all the call sites when adding a new param with a reasonable default.

**Review by jpbetz — CHANGES_REQUESTED** [review]

File: etcdserver/api/membership/cluster_test.go (hunk 1 above)
Comment (jingyih) [reply etcdserver/api/membership/cluster_test.go]: This existing test does not explicitly set isLearner field, and the field is defaulted to false. It is a good idea to have a test with isLearner absent, set to false, and set to true. I will add it.

File: etcdserver/api/membership/member.go (hunk 4 above)
Comment (jingyih) [reply etcdserver/api/membership/member.go:56]: Agreed.

File: etcdserver/api/membership/member.go (hunk 3 above)
Comment (jingyih) [reply etcdserver/api/membership/member.go]: Good catch. I think it is safer with "omitempty".

File: etcdserver/api/membership/cluster.go (hunk 2 above)
Comment (jingyih) [reply etcdserver/api/membership/cluster.go]: Done.

**Comment by gyuho** [comment]
Good first change. Thanks for working on this. LGTM.

**Comment by jpbetz** [comment]
LGTM

**Pull request merged**